### PR TITLE
[android] Move the string and other tags in pointers to the second byte because Android enabled memory tagging

### PR DESCRIPTION
**Pull request by finagolfin**
Android added [Software Tagging (ST from here on) to the top byte in their malloc a couple years ago](https://android.googlesource.com/platform/bionic/+/3b21ada%5E%21/), which has [rolled out to some fraction of Android 11+ devices now and collides with Swift's String tag in the top half of the top byte](https://forums.swift.org/t/swift-runtime-error-running-on-android-sdk-30-arm64-v8a/45457). While the software tagging is in preparation for switching to [hardware tagging with ARM's MTE](https://community.arm.com/arm-community-blogs/b/architectures-and-processors-blog/posts/enhanced-security-through-mte), for some reason [ST sets a tag across the entire top byte](https://android.googlesource.com/platform/bionic/+/master/libc/bionic/malloc_tagged_pointers.h#39) even though MTE is only supposed to use the bottom half of the top byte.

I was surprised to find that @milseman had reorganized the String tag to accommodate MTE in #21310 three years ago, which was followed by #21453, but since ST writes to the top half too, that doesn't work. Specifically, ST causes two problems with Swift:

1. Every large string has `B4` written to the top byte when allocated by `malloc`, which means [it's tagged as immortal, small, and foreign](https://github.com/apple/swift/blob/0149ccd0ca69a8e97a4d13d172c8ce67f5498a60/stdlib/public/core/StringObject.swift#L43), with the last one causing errors on Android. Running the Swift compiler validation suite, about 650 test modules have failures, with [this earlier patch of mine that _only_ addresses this String issue by moving the small and foreign bits fixing half of those failing tests](https://gist.github.com/buttaface/a7fc5b91db4f8068a8ce1af89dbfe355), though it means ARC is not run on all those "immortal" strings and they leak.
2. Every malloc'ed object, like arrays or large strings, has that tag set, which means ARC sees them all as immortal because the top bit is set. After applying this pull, which enables ARC by also moving that immortal top bit, I see a bunch of segfaults when deallocating, because ARC now removes the ST tag somewhere and Bionic complains that the ST tag has been removed.

@hctim, you added this software tagging a couple years ago: any reason why you didn't confine it to the bottom half of the top byte, as the hardware MTE is supposed to do? If Android sticks with MTE in the future, this pull to work around the current ST won't be necessary.

I'm asking before I go any further to find where those tags are being removed, and slapping this pull up as a draft for feedback in the meantime.

### CODE REVIEW

File: stdlib/public/core/Builtin.swift
Comment (buttaface) [inline stdlib/public/core/Builtin.swift:402]: I don't think this should be used because ObjC interoperability is disabled on Android, same for `largeIsCocoa()` and `largeCocoa()` below, but just played it safe by shifting these too, as some code may still call these somewhere.

**Comment by buttaface** [comment]
I asked about Android's new tagging in android/ndk#1653 earlier this month and was told the OS reserves the entire top byte of pointers since Android 11, so I reworked this pull to move Swift's tags to the second byte. This pull now gets almost all the same tests to pass in the compiler validation suite as on a device without tagging, with only four more test modules failing on an Android device with tagging enabled. [One is expected to fail because of the moved tag](https://github.com/apple/swift/pull/21453/files#diff-2f6f1b49b8d1214fab4f0bfdf2126cf167562b7d1c69f4068f6da523a7caf83fR38), and the other three don't fail if I run the test in-process and a few test cases at a time, so there's something flaky in `StdlibUnittest` there.

However, once I build Foundation with this pull and try to run its tests on a device with tagging, I get a segfault when trying [to instantiate a testcase array and append to it](https://github.com/apple/swift-corelibs-foundation/blob/main/Tests/Foundation/Tests/TestTimeZone.swift#L292), in the compiler-generated function `__swift_instantiateConcreteTypeFromMangledNameAbstract` when [it checks to see if the address is negative](https://github.com/apple/swift/blob/e7e536705ee98b917049d64c7d3c973225ef3b59/lib/IRGen/MetadataRequest.cpp#L2747), which is now true even for a valid address because Android's tag always sets the top bit. I will look into this more, examining #26455 where that function was added now.

@milseman, would you review what I have so far?

**Comment by milseman** [comment]
How are you testing these changes on Android?

File: stdlib/public/SwiftShims/HeapObject.h
Comment (milseman) [inline stdlib/public/SwiftShims/HeapObject.h:166]: Should we make a separate `SWIFT_ABI_DEFAULT_BRIDGEOBJECT_TAG_64` for Android? Something like a `SWIFT_ABI_ANDROID_BRIDGEOBJECT_TAG_64`?

File: stdlib/public/core/Builtin.swift
Comment (milseman) [reply stdlib/public/core/Builtin.swift:402]: Is it possible to phrase that in terms of an assertion (i.e. `_invariantCheck`)? That way it'll at least show up when testing.

File: stdlib/public/core/KeyPath.swift
Comment (milseman) [inline stdlib/public/core/KeyPath.swift:1769]: Someone else should review this impact. We may want/need to tidy up our macros and builtins some to avoid a lot of error-prone `#if` logic.

File: stdlib/public/core/KeyPath.swift
Comment (milseman) [inline stdlib/public/core/KeyPath.swift:3252]: Same.

File: stdlib/public/runtime/HeapObject.cpp
Comment (milseman) [inline stdlib/public/runtime/HeapObject.cpp:72]: Same, someone else should be commenting on the impact here and whether we could clean up how we define our ABI some.

**Review by milseman — COMMENTED** [review]
This seems like the correct way to go about implementing this direction, modulo some cleanups, if this is the right direction to go.

What's the testing story on Android?

File: stdlib/public/core/KeyPath.swift
Comment (buttaface) [reply stdlib/public/core/KeyPath.swift:1769]: @jckarter, you added these alignment checks to this file in #10419, please let me know what you think of these two changes to avoid pointer tags.

File: stdlib/public/runtime/HeapObject.cpp
Comment (buttaface) [reply stdlib/public/runtime/HeapObject.cpp:72]: @jckarter, you added the negative check below in #7732: let me know if this is a good way to do this, given that Android AArch64 reserves the top byte and I moved the immortal bit to the top of the second byte.

**Comment by buttaface** [comment]
I put together [a small patch for the metadata issue](https://github.com/buttaface/termux-packages/commit/bf4b23fd0929b0f2488a43df20c355ad445bb484#diff-ffe215fbfee535154ee1814cfe4b5e1024f4d3ddbd90058fe4034232e34619afR9) and it got the remaining parts of the trunk toolchain building and passing their tests up through SPM, plus the three flaky test modules from the validation suite that I mentioned above all pass now. It is a hack as it just checks for the static tag `B4` that Android uses now, but that tag will change in the future so this is not a long-term solution. Still, I've kicked off a build of Swift 5.5.2 with that patch, to see if it works now.

File: stdlib/public/SwiftShims/HeapObject.h
Comment (buttaface) [reply stdlib/public/SwiftShims/HeapObject.h:166]: We could, but note that this is only shifted for AArch64, so I'd have to define it only for that. I figured shifting it here is the best way to set that up, but I could add another constant instead if you like.

File: stdlib/public/core/Builtin.swift
Comment (buttaface) [reply stdlib/public/core/Builtin.swift:402]: I'll look into it.

File: lib/IRGen/MetadataRequest.cpp
Comment (buttaface) [inline lib/IRGen/MetadataRequest.cpp:2771]: Reworked the metadata patch I linked before and have it working for an arbitrary Android tag in the top byte, by using the scheme described in the comments.

Strictly speaking, only the 24th bit matters on Android AArch64 with this new scheme, but I have it checking the 32nd "sign bit" too.

This pull now gets trunk building and passing all but [the one test I linked before](https://github.com/apple/swift/pull/21453/files#diff-2f6f1b49b8d1214fab4f0bfdf2126cf167562b7d1c69f4068f6da523a7caf83fR38), both on devices with and without pointer tagging.

File: lib/IRGen/SwiftTargetInfo.cpp
Comment (buttaface) [inline lib/IRGen/SwiftTargetInfo.cpp:41]: Updated to use the right tag mask at runtime, as selecting it at compile-time as I had it before wouldn't apply the right mask when cross-compiling, and rebased.

File: stdlib/public/SwiftShims/System.h
Comment (buttaface) [inline stdlib/public/SwiftShims/System.h:158]: I only moved the spare bits mask initially, as this ObjC mask appeared to have no effect, but after turning stdlib assertions on to test the new `_internalInvariantFailure` assertions below, this mask being in the top byte caused several tests to fail, because of the `isNative` checks some assertions added. Because of those stdlib assertion failures alone, I shifted this ObjC mask and all stdlib tests then passed in Debug+assert mode.

File: stdlib/public/core/Builtin.swift
Comment (buttaface) [reply stdlib/public/core/Builtin.swift:403]: I checked and this function and `largeCocoa()` are unused when running the compiler validation suite but almost 200 stdlib tests do check `largeIsCocoa()`, so I added these `_internalInvariantFailure()` traps for these two functions that are unused on Android AArch64. That way, everything is shifted properly when assertions are turned off, but these two functions will fail when assertions are turned on in Android AArch64, to warn anyone invoking them not to use them.

**Comment by buttaface** [comment]
@milseman, this pull now fixes all the tagging collisions on Android AArch64 and I responded to all your review comments. Let me know what you think.

**Comment by buttaface** [comment]
@milseman, anything holding this up?

@artemcm, would you run the CI?

**Comment by buttaface** [comment]
@CodaFi, this pull has been ready for three weeks now, but I can't get anyone to respond anymore. Maybe you can take over the review? Since everything is scoped to Android AArch64 only, this should be perfectly safe to merge.

File: lib/IRGen/MetadataRequest.cpp
Comment (jckarter) [inline lib/IRGen/MetadataRequest.cpp:2686]: Is the memory tag always nonzero? Could we have a scheme where the relative address to the mangled string has `00` in the top byte, and any tag in the top byte indicates the metadata pointer has been resolved?

File: stdlib/public/core/KeyPath.swift
Comment (jckarter) [reply stdlib/public/core/KeyPath.swift:1769]: Masking off the top bytes should not affect the alignment computation. Is the problem here the signed-ness of `%`? How about one of these instead, which are also probably more accurate on other platforms:

```
var misalign = Int(UInt(bitPattern: baseAddress)) % alignment
```

or

```
var misalign = Int(bitPattern: baseAddress) & (alignment - 1)
```

Comment (jckarter) [reply stdlib/public/core/KeyPath.swift:1769]: Or if the standard library has since grown standard APIs for pointer alignment, maybe we can adopt those here.

File: stdlib/public/runtime/HeapObject.cpp
Comment (jckarter) [reply stdlib/public/runtime/HeapObject.cpp:72]: As you've written it, this is undefined behavior if tag bytes are shifted out because of signed overflow. However, the goal on other 64-bit platforms here is to reserve some address space for "free" for small closure contexts and such, since checking for `x == 0` and `x < 0` are both one CPU instruction, but I don't think you can do the comparison you'd need here without putting another instruction on the `swift_retain` hot path. We don't currently take advantage of this optimization, so you could do `p != nullptr` like on 32-bit platforms for now.

File: lib/IRGen/MetadataRequest.cpp
Comment (milseman) [reply lib/IRGen/MetadataRequest.cpp:2686]: @buttaface?

File: stdlib/public/core/KeyPath.swift
Comment (milseman) [reply stdlib/public/core/KeyPath.swift:1769]: @buttaface ?

File: stdlib/public/runtime/HeapObject.cpp
Comment (milseman) [reply stdlib/public/runtime/HeapObject.cpp:72]: @buttaface ?

File: stdlib/public/core/KeyPath.swift
Comment (buttaface) [reply stdlib/public/core/KeyPath.swift:1769]: > Is the problem here the signed-ness of %?

Probably so, the memory tag on Android could be anything and right now it is always statically set to `B4` in the top byte, which was changing the result of this modulo calculation.

I'll try your alternatives and report back, don't know what methods the stdlib provides.

File: stdlib/public/runtime/HeapObject.cpp
Comment (buttaface) [reply stdlib/public/runtime/HeapObject.cpp:72]: > this is undefined behavior if tag bytes are shifted out because of signed overflow

I'll look into it, and try your alternative.

File: lib/IRGen/MetadataRequest.cpp
Comment (buttaface) [reply lib/IRGen/MetadataRequest.cpp:2686]: > Is the memory tag always nonzero? 

I don't think we can assume it is anything, it is simply reserved for the OS now.

> Could we have a scheme where the relative address to the mangled string has 00 in the top byte, and any tag in the top byte indicates the metadata pointer has been resolved?

I don't think it would be good to assume anything about the top byte, even though you may be right about what tagging scheme they end up using (currently it's just a static tag, but [they plan to use ARM's hardware MTE and some new scheme they come up with](https://github.com/android/ndk/issues/1653#issuecomment-1011378779)).

I only continue to check the top bit because it was easier to `AND` the new Android check for the eighth bit below than check one `OR` the other, and the result is the same.

I think this is better than trying to unify the scheme for all platforms, _unless_ you plan on turning on MTE for Apple platforms too, in which case this will eventually be needed there too.

File: stdlib/public/core/KeyPath.swift
Comment (buttaface) [reply stdlib/public/core/KeyPath.swift:1769]: > `var misalign = Int(UInt(bitPattern: baseAddress)) % alignment`

I just tried it and six stdlib test modules fail, presumably because this doesn't avoid the problem of the top sign bit being set.

> `var misalign = Int(bitPattern: baseAddress) & (alignment - 1)`

All the same stdlib tests pass when built in release mode, but this assumes the alignment is a power of two. I figured you were using the modulo before because you couldn't assume that, can you?

I ran some searches in the stdlib and while there is a `Builtin._roundUp()` function, I don't see anything that returns the misalignment, as we're calculating here.

File: stdlib/public/runtime/HeapObject.cpp
Comment (buttaface) [reply stdlib/public/runtime/HeapObject.cpp:72]: > `p != nullptr`

This doesn't work, presumably because it never checks for the immortal bit, ie what you check on 64-bit platforms with `p < 0`.

I looked into why the left shift is considered undefined behavior, and as you noted it's because of overflow, but that only matters if you care about doubling the number by shifting, which we don't care about here.

The aim here is simply to check if the ninth bit is set and if the pointer is non-zero if that immortal bit isn't set, ie the same checks as the 64-bit platforms below but simply with an immortal bit located a byte down, so I think this left shift I used should work fine.

File: stdlib/public/core/Builtin.swift
Comment (buttaface) [reply stdlib/public/core/Builtin.swift:409]: > Should we just have an unreachable statement instead?

I'm fine with anything for these two unused methods `_objectPointerIsObjCBit` and `largeCocoa()` at this point. I can leave them untouched, trap, or apply any other mixed approach like this that you want.

I had initially modified these two methods only because I didn't know if they were used somewhere, but now that a full run of the compiler validation suite on a debug+asserts build of the stdlib for Android AArch64 passed, I'm reasonably confident these two are completely unused and can revert all changes to them if you prefer.

Just let me know what you think is best.

File: stdlib/public/SwiftShims/System.h
Comment (milseman) [reply stdlib/public/SwiftShims/System.h:158]: What happens if we just omit this definition then? That is, does it have less source impact to omit objc masks and support?

File: stdlib/public/core/KeyPath.swift
Comment (milseman) [reply stdlib/public/core/KeyPath.swift:1769]: @jckarter I couldn't tell from the conversation, but has your concern been addressed?

File: stdlib/public/runtime/HeapObject.cpp
Comment (milseman) [reply stdlib/public/runtime/HeapObject.cpp:72]: @buttaface can you address the undefined behavior aspect?

**Review by milseman — COMMENTED** [review]
Overall this is LGTM, though I'd like to minimize the number of source changes to areas that aren't supported anyways (ObjC masks and bits, etc).

File: stdlib/public/core/Builtin.swift
Comment (buttaface) [reply stdlib/public/core/Builtin.swift:409]: I will revert all my changes to these two ObjC-related methods.

File: stdlib/public/SwiftShims/System.h
Comment (buttaface) [reply stdlib/public/SwiftShims/System.h:158]: This OBJC mask is completely unused except in some assertions, which simply check `isNative` in some invariants, even though that makes no sense with ObjC interoperability turned off? Rather than have all those tests fail when assertions are turned on, I added this mask.

However, I don't really care if this assertions workaround is upstreamed and can remove it if you like.

File: stdlib/public/runtime/HeapObject.cpp
Comment (milseman) [reply stdlib/public/runtime/HeapObject.cpp:72]: Unfortunately, that's not what undefined behavior means. I'm not concerned UB might produce the wrong results, because that would mean we actually got lucky.

You can use an unsigned representation to make it defined or otherwise work around the issue

Comment (buttaface) [reply stdlib/public/runtime/HeapObject.cpp:72]: OK, I probably don't understand fully what UB means in the context of this C++ then. My understanding is that left shifts are defined on signed integers too, but they may overflow if the input number to be shifted gets too large, just like with unsigned integers.

Since in this case, I know exactly what the input is, ie a tag in the top byte that I _want_ shifted out, and an immortal bit in the ninth bit that I then want to run the same `> 0` check on, I don't see this as UB. Changing this to unsigned would defeat the purpose, which is to use the same `> 0` check as used for other 64-bit platforms a couple lines below.

File: stdlib/public/SwiftShims/System.h
Comment (buttaface) [reply stdlib/public/SwiftShims/System.h:158]: The problem is that without this change [the default is to reserve the top bit on all AArch64 platforms](https://github.com/apple/swift/blob/0ab3eec987aefadf7dd97740d7baaa785005edb9/stdlib/public/SwiftShims/System.h#L161), which is then [checked against the top bit in Android's memory tag in `isNative`](https://github.com/apple/swift/blob/abec55f4324c98ef69b94100b649c83bdbfc8f91/stdlib/public/core/BridgeStorage.swift#L89), and `isNative` is checked by several assertions.

Presumably this is a mistake in checking this ObjC bit even when ObjC interoperability is turned off, so I worked around it by moving the ObjC bit down a byte, out of the Android memory tag in the top byte.

If you'd rather set this mask to zero everywhere on Android AArch64, that might work too.

File: stdlib/public/core/Builtin.swift
Comment (buttaface) [reply stdlib/public/core/Builtin.swift]: Done.

File: stdlib/public/runtime/HeapObject.cpp
Comment (buttaface) [reply stdlib/public/runtime/HeapObject.cpp]: I looked into this more and apparently [there is a nonsense declaration in some recent C++ versions](https://en.cppreference.com/w/cpp/language/operator_arithmetic) that `For negative a, the behavior of a << b is undefined`. Since this can only be enforced if `a` is known at compile-time or by inserting a runtime check to see if `a` is negative before shifting, [Regehr notes it is never enforced as it would break a ton of code](https://blog.regehr.org/archives/738).

My intent here was to keep this quick check for this commonly invoked function and the shifted version compiles down to the following tight three instructions, after I disabled the mandatory inlining:
```
       0: 08 dc 78 d3  	lsl	x8, x0, #8
       4: 1f 01 00 f1  	cmp	x8, #0                  // =0
       8: e0 d7 9f 1a  	cset	w0, gt
       c: c0 03 5f d6  	ret
```
I then wrote the equivalent by masking and comparing twice, with no shifts, pushing it up to five instructions:
```
       0: 1f 00 49 f2  	tst	x0, #0x80000000000000
       4: e8 17 9f 1a  	cset	w8, eq
       8: 1f dc 40 f2  	tst	x0, #0xffffffffffffff
       c: e9 07 9f 1a  	cset	w9, ne
      10: 00 01 09 0a  	and	w0, w8, w9
      14: c0 03 5f d6  	ret
```
I'm skeptical that any C++ compiler would ever dare break this commonly used idiom, so I think we should just use the shifted version, which I've commented out in this pull for now. If you don't want to do that, I'm fine with the longer mask and compares version instead.

@drodriguez, you may want to chime on this and the whole pull.

Comment (drodriguez) [reply stdlib/public/runtime/HeapObject.cpp]: Not my area of expertise, but I agree with Michael and Joe that using undefined behaviour because it works for your use case today is not the way to go.

I was trying  to understand what we were trying, and it seems that you want to move the bit 56 into the sign bit of the 64 bit register (assuming two's complement), but I think you can also go the other way:

```
return (int32_t)((intptr_t)p >> 16) > 0;
```

This end up with 3 instructions in Godbolt with -O3, which put it on par to the solution that relies on UB:

```
        lsr     x8, x0, #16
        cmp     w8, #0                          // =0
        cset    w0, gt
        ret
```

Double-check that my idea is solid and return the same results, and wait to see if anyone else can find the hole in my proposal.

PD: I skimmed the rest of the proposal, and I think there's some cases where you write comments about the 24th bit, and I think you meant the 56th bit.

Comment (buttaface) [reply stdlib/public/runtime/HeapObject.cpp]: > it seems that you want to move the bit 56 into the sign bit of the 64 bit register

That's part of it: as Joe noted above, we need to know if the immortal bit is set, `x < 0` normally, and that the pointer is non-null, `x != 0`. Your alternate formulation is throwing away information for the second check and will only work if you're sure that no pointer will only be contained in the bottom 16 bits.

My most recent mask and compares version in this pull checks both completely.

> there's some cases where you write comments about the 24th bit, and I think you meant the 56th bit.

I think you're referring to the metadata changes alone. There, the compiler places two 32-bit words into a 64-bit pointer in the uninitialized cache: one containing an Int32 with the string's size and the other with a 32-bit relative address to the mangled string. So in that case, my comments about the 24th bit only refers to that first 32-bit word with the size, since that's the only one that's checked, even though we check it with a 64-bit mask to the entire pointer.

Comment (drodriguez) [reply stdlib/public/runtime/HeapObject.cpp]: > Your alternate formulation is throwing away information for the second check

I was trying to reproduce your initial code (`(intptr_t)p << 8 > 0`) but instead of moving `<<`, moving `>>` and leaving the 56th bit on the 32th place and using a 32 bit comparison for positive (maybe the offset should be 24 and not 16 in my code?). Was your initial code also not having into account the second check?

> the compiler places two 32-bit words

That makes sense. As I said, I just skimmed, and the discussion talking about 24th was unexpected.

Comment (buttaface) [reply stdlib/public/runtime/HeapObject.cpp]: > maybe the offset should be 24 and not 16 in my code?

Yep, should be 24.

> Was your initial code also not having into account the second check?

Let me back up a bit and explain what we're doing here. This inlined function `isValidPointerForNativeRetain()` is used throughout the ARC functions in this file to determine if ARC should run or not. It does so by checking if the pointer has the immortal bit set, which on 64-bit platforms is the top "sign" bit as can be seen elsewhere in this pull, or if it is a null pointer, and it ignores the pointer in both cases.

On Android AArch64 with memory tagging, the top byte of these pointers is reserved for the OS, so Swift's immortal bit can no longer be in the top bit, and I move it one byte down to the ninth bit in this pull. The shifted version of this check tries to preserve that same quick `p > 0` comparison by shifting the immortal bit back into the top sign bit and a shifted null pointer is still a null pointer.

What I have in this pull now by masking and comparing is the explicit version of the same check:
```
((0x0080000000000000ULL & (uintptr_t)p) == 0) && ((0x00FFFFFFFFFFFFFFULL & (uintptr_t)p) > 0)
```
I think we should use the quicker shifted version, but I'm fine with this instead.

**Comment by buttaface** [comment]
@milseman, been a couple weeks now, can we go ahead and get this in, as I don't want to miss the 5.7 branch? If you can just sign off on [the two](https://github.com/apple/swift/pull/40779#discussion_r796334743) unresolved [issues you raised](https://github.com/apple/swift/pull/40779#discussion_r826283157), and either you decide on the three we're waiting on Joe for or we can just get this in and always change it later whenever he chimes in, since this pull doesn't affect any officially supported platform, only Android AArch64. 

I will rebase and update this pull once you let me know on those review comments.

File: stdlib/public/runtime/HeapObject.cpp
Comment (jckarter) [reply stdlib/public/runtime/HeapObject.cpp]: If you want to use the shifted version, then you can use unsigned rather than signed types. But I don't think any of this is necessary—it's sufficient and cheaper to check for `!= 0`. We do not currently use the negative bit optimization for anything on existing 64-bit platforms, and when we do, we would need to disable or modify those optimizations for MTE-enabled Android anyway, since there would be fewer bits available for storing data than on other platforms.

File: stdlib/public/core/KeyPath.swift
Comment (jckarter) [reply stdlib/public/core/KeyPath.swift:1769]: We can assume the alignment is a power of two. We should use `var misalign = Int(bitPattern: baseAddress) & (alignment - 1)` everywhere. The existing code would also be a bug on other 64-bit platforms that might have the high bit set for a valid pointer.

**Comment by jckarter** [comment]
It would be good to avoid introducing UB, fix bugs, and avoid unnecessary conditional compilation, so if we can get the fix to the alignment computation in KeyPath, and remove the UB in the null-or-negative pointer check, then I think we're OK.

File: stdlib/public/core/KeyPath.swift
Comment (buttaface) [reply stdlib/public/core/KeyPath.swift:1769]: Will do.

File: stdlib/public/runtime/HeapObject.cpp
Comment (buttaface) [reply stdlib/public/runtime/HeapObject.cpp]: > If you want to use the shifted version, then you can use unsigned rather than signed types.

I'll try that: I'll see if clang accepts shifting it as an unsigned type, then casting the result to a signed type for the comparison.

> But I don't think any of this is necessary—it's sufficient and cheaper to check for != 0. We do not currently use the negative bit optimization for anything on existing 64-bit platforms, and when we do, we would need to disable or modify those optimizations for MTE-enabled Android anyway, since there would be fewer bits available for storing data than on other platforms.

You're saying the top bit is currently never intentionally set for pointers as an immortal bit on 64-bit platforms? Or that we should now ignore the immortal bit when set on Android AArch64? Because when I tried your `p != nullptr` alternative earlier, several stdlib tests started failing, whereas my two alternatives here that check the immortal bit keep them passing.

I can investigate further why that is, if you believe only checking for null pointers should suffice.

Comment (jckarter) [reply stdlib/public/runtime/HeapObject.cpp]: Hm, which tests are those that fail?

**Comment by milseman** [comment]
I agree that we don't want to add UB to the swift runtime

File: stdlib/public/runtime/HeapObject.cpp
Comment (buttaface) [reply stdlib/public/runtime/HeapObject.cpp]: I ran a little experiment to see if I can reproduce the stdlib test failures on linux x86_64 and I can. I setup a Ubuntu 20.04 VPS and downloaded the latest Mar. 30 trunk snapshot build and source, before running the following command to build the standalone stdlib and run its tests:
```
./swift/utils/build-script -RA --build-swift-tools=0 --skip-build-cmark --build-llvm=0
--native-clang-tools-path=/home/butta/swift-DEVELOPMENT-SNAPSHOT-2022-03-30-a-ubuntu20.04/usr/bin
--native-swift-tools-path=/home/butta/swift-DEVELOPMENT-SNAPSHOT-2022-03-30-a-ubuntu20.04/usr/bin
-j9 -T --lit-args=--filter=stdlib/ --skip-test-early-swift-driver
```
That runs about 530 stdlib tests and only 5 fail, because I don't have needed utilities like llvm-nm installed.

Then I changed the line below from `return (intptr_t)p > 0` to `return p != nullptr` and reran the above build command, only to see 350 tests failing now. Looking at the results, I see several segfaults like this, ie many test binaries segfault if we only check for the null pointer:
```
/home/butta/build/Ninja-Release/swift-linux-x86_64/validation-test-linux-x86_64/stdlib/Output/Bitset.swift.script:
line 1: 356022 Segmentation fault      (core dumped)
/usr/bin/env LD_LIBRARY_PATH='/home/butta/build/Ninja-Release/swift-linux-x86_64/lib/swift/linux'
/home/butta/build/Ninja-Release/swift-linux-x86_64/validation-test-linux-x86_64/stdlib/Output/Bitset.swift.tmp/a.out
```
Around the same number of stdlib tests failed on Android AArch64 when I only checked for the null pointer a couple weeks ago as you suggested, which you can reproduce on linux if you like using the above steps: let me know if it is unexpected and if we should investigate further.

File: stdlib/public/core/KeyPath.swift
Comment (buttaface) [reply stdlib/public/core/KeyPath.swift]: Done.

File: stdlib/public/runtime/HeapObject.cpp
Comment (buttaface) [reply stdlib/public/runtime/HeapObject.cpp]: I've updated this line to shift again with `return (intptr_t)((uintptr_t)p << 8) > 0`, which maintains the same shifting three-instruction assembly output above and hopefully threads your UB needle too.

File: stdlib/public/SwiftShims/System.h
Comment (buttaface) [reply stdlib/public/SwiftShims/System.h]: I set this OBJC mask to zero and ran the full compiler validation suite on a Debug+asserts stdlib and got the same passing test results, so I've now set this mask to zero. Since this is just a workaround for stdlib assertions that wrongly check `isNative` even when ObjC interop is turned off, I'm fine with not modifying this OBJC mask in any way in this pull and fixing that root cause in a separate pull later instead.

@milseman, just let me know which you prefer.

File: stdlib/public/core/KeyPath.swift
Comment (buttaface) [reply stdlib/public/core/KeyPath.swift]: Updated, same as above.

**Comment by buttaface** [comment]
@milseman and @jckarter, if you would take a look at the latest iteration for those three remaining issues, we could finally run this through the CI and get it merged.

**Review by milseman — APPROVED** [review]
LGTM, assuming the runtime stuff looks good and no UB

**Comment by jckarter** [comment]
Looks fine to me now.

**Comment by buttaface** [comment]
Rebased and updated a comment, @artemcm, would you run the CI?

**Comment by buttaface** [comment]
@kateinoigakukun, I see you're up, mind running the CI on this pull for me?

**Comment by kateinoigakukun** [comment]
@swift-ci Please test

**Comment by buttaface** [comment]
@milseman, passed CI, would you merge?

**Pull request merged**
